[FLINK-12628][Runtime / Coordination] Check test failure if partitionhas no consumers in Execution.getPartitionMaxParallelism

liyafan82 opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Travis CI to do that following [this guide](http://flink.apache.org/contribute-code.html#best-practices).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

Resolve issue FLINK-12628 Check test failure if partition has no consumers in Execution.getPartitionMaxParallelism:

Currently, we work around this case in Execution.getPartitionMaxParallelism because of tests:

// TODO consumers.isEmpty() only exists for test, currently there has to be exactly one consumer in real jobs!

though partition is supposed to have always at least one consumer atm.
We should check which test fails and consider fixing it.

According to my investigation, there is no test failure, when we ignore the case for consumers.isEmpty() equals to true.

## Brief change log

  - Change the implementation of Execution.getPartitionMaxParallelism to ignore the case for consumers.isEmpty() equals to true.


## Verifying this change

This change is already covered by existing tests, such as ExecutionTest.


## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

klion26 commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/Execution.java:
```diff
@@ -679,14 +678,10 @@ public void setInitialState(@Nullable JobManagerTaskRestore taskRestore) {
-			ExecutionJobVertex consumerVertex = consumer.get(0).getTarget().getJobVertex();
-			maxParallelism = consumerVertex.getMaxParallelism();
-		}
+		List<ExecutionEdge> consumer = consumers.get(0);
```
As the previous comment said, `consumers will be empty in test`, so `consumers.get(0)` will throw an exception here?

liyafan82 replied:
Maybe not. In my local environment, no exception was thrown when we ignore the case for the consumer being empty. Let's see if we can reproduce it in Travis.

azagrebin replied:
We can also add a check if this assumption holds to better debug it if fails:
`Preconditions.checkArgument(!consumers.isEmpty(), "Currently there has to be exactly one consumer in real jobs");`

azagrebin commented:
@liyafan82 
I rebased the PR on the latest master and added the precondition check, I mentioned in the comment.
Do you have time to work on this?
We moved recently to other in house Travis CI. It looks like it is triggered only for new PRs. We could close this PR and open a new one to see Travis results.

liyafan82 commented:
> @liyafan82
> I rebased the PR on the latest master and added the precondition check, I mentioned in the comment.
> Do you have time to work on this?
> We moved recently to other in house Travis CI. It looks like it is triggered only for new PRs. We could close this PR and open a new one to see Travis results.

Sure. I will start a new PR for this. Thanks.

azagrebin commented:
@liyafan82 
the CI actually reported success, we can continue with this PR so I will reopen it, sorry for confusion.

liyafan82 commented:
> @liyafan82
> the CI actually reported success, we can continue with this PR so I will reopen it, sorry for confusion.

Sure. Thanks for your effort.
